documentation/specs/proposed/General_perf_onepager.md

```diff
@@ -0,0 +1,42 @@
+# General performance improvements
+MSBuild is the main tool used for building .NET projects. It is used by Visual Studio, the .NET CLI, and other build tools. The performance of MSBuild is critical to the productivity of developers. This document outlines our goals to improve overall performance of MSBuild.
+
+## Goals and Motivation
+
+We are aiming for:
+ - Searching opportunities to improve the build times.
+ - Reducing MSBuild overhead in the Visual Studio IDE.
+ - Collecting data to understand the current performance bottlenecks.
+
+This document does not cover specific performance improvements, but rather outlines the general goals and motivation for the performance improvements.
+
+
+## Impact
+    
+    - Opening project/solution, branch switching and other operations using MSBuild code in VS should be less impacted by the MSBuild.
+    - Overall build times should be reduced. 
+    - Even a small improvement can save a lot of time and computing resources across all builds done daily worldwide.
+
+## Stakeholders
+
+    - Chet Hust (PM) - as a customer advocate
+    - David Kean - as a VS performance expert
+
+## Risks
+
+    - Performance improvements might not be as significant as expected.
+    - We can break existing functionality while making changes.
+    - We can thow away some ideas due to technical limitations or complexity/improvements ratio.
+
+## Cost
+
+Performance improvements are a long-term investment. We need to balance the cost of the improvements with the expected benefits.
+We will need to invest time in collecting data, analyzing the data, and implementing improvements.
+
+Our goal in this scope is to find small and medium size opprotunities. Bigger functionalities such as evaluation caching and RAR caching are described in separate documents.
+
+## Plan
+    
+    - Collect data on the current performance bottlenecks.
```

Comment:
We might want to name some of the strategies to be used:
 * Top hits in PRISM
 * ETW measurements on batch build / incrmental build and VS build / incremental build on small and mid solutions and detecting opportunities to improve
 * Reviewing some of the earlier identified opportunities: https://github.com/dotnet/msbuild/issues?q=is%3Aissue%20state%3Aopen%20label%3Aperformance

Reply:
Visual Studio can 100% help with this, we have:

1) Performance lab traces with representative samples across almost all project types, including C++, .NET Framework and .NET Core.

2) Real world telemetry of our builds in the wild. I would focus on rebuild where we actually call MSBuildas we mostly try and avoid skipping msbuild when we can due to performance reasons.

3) We are already collecting, or can collect real world traces from machines in the wild for:
   - High CPU usage
   - High GC-time
   - Thread pool starvation
   - Solution load (for evaluation time)
   - Builds (just turning this on, this week)
 
These are analyzed, blamed and then aggregated through our analyzers. These produce bug-level failures that that are ranked by the top issues in the wild. From here we can find very easy low hanging fruit across the entire product. It can, or will identify, big ticket items as well, such as RAR, JITing overhead, lack of Server GC (so GC can't compete), CPU starvation + cost of context switches (due to MSBuild firing up more thread pools than CPUs across all nodes). I suggest we start with the easy, and low hanging fruit and then build the case for big ticket items. I can help with this.

Once we have data, I will come and present and teach you folks how to read this data, download traces, understand the data, etc.